url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/optimization/diffevo/DifferentialEvolution.java

```diff
@@ -0,0 +1,607 @@
+        //
+        int offset_a = this.offset(a);
+        int offset_b = this.offset(b);
+        int offset_c = this.offset(c);
+        //
+        final int n = this.getParameters();
+        //
+        for (int j = 0; j < n; j++) {
+            //
+            this.accu[j] = (
+                this.population[offset_a] + 
+                (
+                    this.F * (
+                        this.population[offset_b] -
+                        this.population[offset_c]
+                    )
+                )
+            );
+            //
+            offset_a++;
+            offset_b++;
+            offset_c++;
+        }
+    }
+
+    private void mutateBestOne(final int i) {
+        this.preventRandomIndex(i);
+        //
+        // pick three indices randomly.
+        //
+        final int best = this.best;
+        this.preventRandomIndex(best);
+        final int a = this.nextRandomIndex();
+        final int b = this.nextRandomIndex();
+        //
+        int offset_best = this.offset(best);
+        int offset_a    = this.offset(a);
+        int offset_b    = this.offset(b);
+        //
+        final int n = this.getParameters();
+        //
+        for (int j = 0; j < n; j++) {
+            //
+            this.accu[j] = (
+                this.population[offset_best] + 
+                (
+                    this.F * (
+                        this.population[offset_a] -
+                        this.population[offset_b]
+                    )
+                )
+            );
+            //
+            offset_best++;
+            offset_a++;
+            offset_b++;
+        }
+    }
+
+    private void mutateRand2BestOne(final int i) {
+        this.preventRandomIndex(i);
+        //
+        // pick three indices randomly.
+        //
+        final int best = this.best;
+        this.preventRandomIndex(best);
+        final int a = this.nextRandomIndex();
+        final int b = this.nextRandomIndex();
+        final int c = this.nextRandomIndex();
+        //
+        int offset_i    = this.offset(i);
+        int offset_best = this.offset(best);
+        int offset_a    = this.offset(a);
+        int offset_b    = this.offset(b);
+        int offset_c    = this.offset(c);
+        //
+        final int n = this.getParameters();
+        //
+        for (int j = 0; j < n; j++) {
+            //
+            this.accu[j] = (
+                this.population[offset_i] + (
+                    this.F * (
+                        this.population[offset_best] -
+                        this.population[offset_a]
+                    )
+                ) + (
+                    this.F2 * (
+                        this.population[offset_b] -
+                        this.population[offset_c]
+                    )
+                )
+            );
+            //
+            offset_i++;
+            offset_best++;
+            offset_a++;
+            offset_b++;
+            offset_c++;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 367:388, 411:432.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.